pkg/internal/webhook/bootstrap/bootstrap.go

```diff
@@ -0,0 +1,290 @@
+	certDir string
+
+	// webhookFileCheckRetryPeriod is the period in which the webhook certificate
+	// and key will be checked. Made variable for testing.
+	webhookFileCheckRetryPeriod time.Duration
+}
+
+// Run will start the Webhook bootstrap operation for policy-approver. Start
+// will block until all needed resources for the policy-approver are made
+// available.  Start will continue to manage these resources in the background.
+func Run(ctx context.Context, opts Options) error {
+	client, err := client.New(opts.RestConfig, client.Options{Scheme: cmpapi.GlobalScheme})
+	if err != nil {
+		return fmt.Errorf("failed to build bootstrapper client: %w", err)
+	}
+
+	b := &bootstrapper{
+		log:                         opts.Log.WithName("webhook").WithName("bootstrapper"),
+		evaluator:                   opts.Evaluator,
+		client:                      client,
+		webhookFileCheckRetryPeriod: time.Second * 2,
+		certDir:                     opts.WebhookCertificatesDir,
+	}
+
+	cache, err := cache.New(opts.RestConfig, cache.Options{
+		Scheme: cmpapi.GlobalScheme, Namespace: "cert-manager",
+		SelectorsByObject: cache.SelectorsByObject{
+			&cmapi.CertificateRequest{}: {Label: labels.SelectorFromSet(labels.Set{"policy.cert-manager.io/webhook": "bootstrap"})},
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to build bootstrapper cache: %w", err)
+	}
+
+	informer, err := cache.GetInformer(ctx, &cmapi.CertificateRequest{})
+	if err != nil {
+		return fmt.Errorf("failed to build bootstrapper informer: %w", err)
+	}
+
+	// queue is used to enqueue CertificateRequests to evaluate approval.
+	queue := make(chan *cmapi.CertificateRequest)
+
+	informer.AddEventHandler(kcache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { queue <- obj.(*cmapi.CertificateRequest).DeepCopy() },
+		UpdateFunc: func(_, obj interface{}) { queue <- obj.(*cmapi.CertificateRequest).DeepCopy() },
+	})
+
+	go func() {
+		if err := cache.Start(ctx); err != nil {
+			b.log.Error(err, "error running bootstrapper cache")
+			return
+		}
+	}()
+
+	if !cache.WaitForCacheSync(ctx) {
+		return errors.New("failed to wait for cache to sync")
+	}
+
+	go func() {
+		defer close(queue)
+
+		for {
+			select {
+			case cr := <-queue:
+				log := b.log.WithValues("namespace", cr.Namespace, "name", cr.Name)
+				log.V(2).Info("received certificaterequest event")
+
+				if err := b.evaluateWebhookCertificateRequest(ctx, log, cr); err != nil {
+					log.Error(err, "failed to evaluate request, re-queueing...")
+					go func() {
+						time.Sleep(time.Second / 2)
+						queue <- cr
+					}()
+				}
+
+			case <-ctx.Done():
+				b.log.Info("shutting down")
+				return
+			}
+		}
+	}()
+
+	return b.waitForWebhookCertificateKey(ctx)
+}
+
+// waitForWebhookCertificateKey is a blocking func which waits for the
+// Webhook's certificate and private key to be available on file.
+func (b *bootstrapper) waitForWebhookCertificateKey(ctx context.Context) error {
```

Comment:
Note: this function doesn't check that the file contains valid, unexpired cert, so I guess there might be a small chance that i.e after a restore & restart this returns before an expired cert has been renewed.